src/neuroglancer/sliceview/compresso/compresso.hpp

```diff
@@ -0,0 +1,593 @@
+
+				if (offset == 0) {
+					if (x == 0) {
+						return 2;
+					}
+					labels[loc] = labels[loc - 1];
+				}
+				else if (offset == 1) {
+					if (x >= sx - 1) {
+						return 3;
+					}
+					labels[loc] = labels[loc + 1];
+				}
+				else if (offset == 2) {
+					if (y == 0) {
+						return 4;
+					}
+					labels[loc] = labels[loc - sx];
+				}
+				else if (offset == 3) {
+					if (y >= sy - 1) {
+						return 5;
+					}
+					labels[loc] = labels[loc + sx];
+				}
+				else if (offset == 4) {
+					if (z == 0) {
+						return 6;
+					}
+					labels[loc] = labels[loc - sxy];
+				}
+				else if (offset == 5) {
+					if (z >= sz - 1) {
+						return 7;
+					}
+					labels[loc] = labels[loc + sxy];
+				}
+				else if (offset == 6) {
+					labels[loc] = locations[index + 1];
+					index++;
+				}
+				else {
+					labels[loc] = offset - 7;
+				}
+				index++;
+			}
+		}
+	}
+
+	return 0;
+}
+
+template <typename LABEL, typename WINDOW>
+int decompress(unsigned char* buffer, size_t num_bytes, LABEL* output = NULL) {
+
+	if (num_bytes < CompressoHeader::header_size) {
+		return 8;
+	}
+	else if (!CompressoHeader::valid_header(buffer)) {
+		return 9;
+	}
+
+	const CompressoHeader header(buffer);
+
+	const size_t sx = header.sx;
+	const size_t sy = header.sy;
+	const size_t sz = header.sz;
+	const size_t voxels = sx * sy * sz;
+	const size_t xstep = header.xstep;
+	const size_t ystep = header.ystep;
+	const size_t zstep = header.zstep;
+
+	if (sx * sy * sz == 0) {
+		return 10;
+	}
+
+	const size_t nx = (sx + xstep - 1) / xstep; // round up
+	const size_t ny = (sy + ystep - 1) / ystep; // round up
+	const size_t nz = (sz + zstep - 1) / zstep; // round up
+	const size_t nblocks = nz * ny * nx;
+
+	size_t window_bytes = (
+		num_bytes 
+			- CompressoHeader::header_size
+			- (header.id_size * sizeof(LABEL))  
+			- (header.value_size * sizeof(WINDOW))
+			- (header.location_size * sizeof(LABEL))
+	);
+	size_t num_condensed_windows = window_bytes / sizeof(WINDOW);
+
+	// allocate memory for all arrays
+	std::vector<LABEL> ids(header.id_size + 1); // +1 to allow vectorized mapping w/ no if statement guarding zero
+	std::vector<WINDOW> window_values(header.value_size);
+	std::vector<LABEL> locations(header.location_size);
+	std::vector<WINDOW> windows(num_condensed_windows);
+
+	size_t iv = CompressoHeader::header_size;
+	for (size_t ix = 0; ix < ids.size() - 1; ix++, iv += sizeof(LABEL)) {
+		ids[ix + 1] = ctoi<LABEL>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < window_values.size(); ix++, iv += sizeof(WINDOW)) {
+		window_values[ix] = ctoi<WINDOW>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < locations.size(); ix++, iv += sizeof(LABEL)) {
+		locations[ix] = ctoi<LABEL>(buffer, iv);
+	}
+	for (size_t ix = 0; ix < num_condensed_windows; ix++, iv += sizeof(WINDOW)) {
+		windows[ix] = ctoi<WINDOW>(buffer, iv);
+	}
+
+	windows = run_length_decode_windows<WINDOW>(windows, nblocks);
+
+	bool* boundaries = decode_boundaries<WINDOW>(
```

Comment:
Use std::unique_ptr<T[]> to hold all of these arrays -- that way you don't need to remember to free manually

Reply:
Done for boundaries, but I am more leery of doing that in cc3d as I am using pointer arithmetic.